salt/config/__init__.py

```diff
@@ -3811,10 +3781,31 @@ def apply_minion_config(overrides=None,
 
     # Check and update TLS/SSL configuration
     _update_ssl_config(opts)
+    _update_discovery_config(opts)
 
     return opts
 
 
+def _update_discovery_config(opts, minion=True):
+    '''
+    Update discovery config for all instances.
+
+    :param opts:
+    :return:
+    '''
+    if opts.get('discovery') not in (None, False):
+        if opts['discovery'] is True:
+            opts['discovery'] = {}
+        discovery_config = {'attempts': 3, 'pause': 5, 'port': 4520, 'match': 'any', 'mapping': {}}
+        for key in opts['discovery'].keys():
+            if key not in discovery_config:
+                raise salt.exceptions.SaltConfigurationError('Unknown discovery option: {0}'.format(key))
+        if not minion:
```

Comment:
We can replace this line with:

```python
if opts.get('__role') == 'master':
```

Reply:
Ah, cool. Didn't know about that. Actually `!= 'minion'`, but point taken. Thanks!